types.go

```diff
@@ -17,72 +17,155 @@ limitations under the License.
 package fission
 
 type (
-	// Metadata is used as the general identifier for all kinds of
-	// resources managed by the controller.
-	Metadata struct {
-		Name string `json:"name"`
-		Uid  string `json:"uid,omitempty"`
+	//
+	// Functions and packages
+	//
+
+	// ChecksumType specifies the checksum algorithm, such as
+	// sha256, used for a checksum.
+	ChecksumType string
+
+	// Checksum of package contents when the contents are stored
+	// outside the Package struct. Type is the checksum algorithm;
+	// "sha256" is the only currently supported one. Sum is hex
+	// encoded.
+	Checksum struct {
+		Type ChecksumType `json:"type"`
+		Sum  string       `json:"sum"`
 	}
 
-	// Function is a unit of executable code.  Though it's called
-	// a function, the code may have more than one function; it's
-	// usually some sort of module or package.
-	Function struct {
-		Metadata    `json:"metadata"`
-		Environment Metadata `json:"environment"`
-		Code        string   `json:"code"`
+	// PackageType is either literal or URL, indicating whether
+	// the package is specified in the Package struct or
+	// externally.
+	PackageType string
+
+	// Package contains or references a collection of source or
+	// binary files.
+	Package struct {
+		// Type defines how the package is specified: literal or URL.
+		Type PackageType `json:"type"`
+
+		// Literal contents of the package. Can be used for
+		// encoding packages below TODO (256KB?) size.
+		Literal []byte `json:"literal"`
+
+		// URL references a package.
+		URL string `json:"url"`
+
+		// Checksum ensures the integrity of packages
+		// refereced by URL. Ignored for literals.
+		Checksum Checksum `json:"checksum"`
+
+		// EntryPoint optionally specifies an entry point in
+		// the package. Each environment defines a default
+		// entry point, but that can be overridden here.
+		EntryPoint string `json:"entrypoint"`
 	}
 
-	// Environment identifies the language and OS specific
-	// resources that a function depends on.  For now this
-	// includes only the function run container image.  Later,
-	// this will also include build containers, as well as support
-	// tools like debuggers, profilers, etc.
-	Environment struct {
-		Metadata             `json:"metadata"`
-		RunContainerImageUrl string `json:"runContainerImageUrl"`
+	// FunctionSpec describes the contents of the function.
+	FunctionSpec struct {
```

Comment:
This is part of the plan for declarative definitions. You can have a yaml file that's the serialized version of these objects, just like any other kubernetes yaml files. (Client tooling to work with those files will be improved in future changes.)